src/datomish/db.cljc

```diff
@@ -12,31 +12,599 @@
+              (:rowid (first (<? (s/all-rows (:sqlite-connection db) ["SELECT last_insert_rowid() AS rowid"]))))))
+          ]
+      rowid)))
+
+(defrecord DB [sqlite-connection schema idents current-tx]
+  ;; idents is map {ident -> entid} of known idents.  See http://docs.datomic.com/identity.html#idents.
   IDB
   (query-context [db] (context/->Context (source/datoms-source db) nil nil))
-  (close [db] (s/close (.-sqlite-connection db))))
 
-(defn <with-sqlite-connection [sqlite-connection]
+  (schema [db] (.-schema db))
+
+  (idents [db] (.-idents db))
+
+  (current-tx
+    [db]
+    (inc (:current-tx db)))
+
+  ;; TODO: use q for searching?  Have q use this for searching for a single pattern?
+  (<eavt [db pattern]
+    (let [[e a v] pattern
+          v (and v (ds/->SQLite schema a v))] ;; We assume e and a are always given.
+      (go-pair
+        (->>
+          {:select [:e :a :v :tx [1 :added]] ;; TODO: generalize columns.
+           :from [:all_datoms]
+           :where (cons :and (map #(vector := %1 %2) [:e :a :v] (take-while (comp not nil?) [e a v])))} ;; Must drop nils.
+          (sql/format)
+
+          (s/all-rows (:sqlite-connection db))
+          (<?)
+
+          (mapv (partial row->Datom (.-schema db))))))) ;; TODO: understand why (schema db) fails.
+
+  (<avet [db pattern]
+    (let [[a v] pattern
+          v (ds/->SQLite schema a v)]
+      (go-pair
+        (->>
+          {:select [:e :a :v :tx [1 :added]] ;; TODO: generalize columns.
+           :from [:all_datoms]
+           :where [:and [:= :a a] [:= :v v] [:= :index_avet 1]]}
+          (sql/format)
+
+          (s/all-rows (:sqlite-connection db))
+          (<?)
+
+          (mapv (partial row->Datom (.-schema db))))))) ;; TODO: understand why (schema db) fails.
+
+  (<apply-datoms [db datoms]
+    (go-pair
+      (let [exec (partial s/execute! (:sqlite-connection db))]
+        ;; TODO: batch insert, batch delete.
+        (doseq [datom datoms]
+          (let [[e a v tx added] datom
+                schema (.-schema db) ;; TODO: understand why (schema db) fails.
+                v (ds/->SQLite schema a v)
+                fulltext? (ds/fulltext? schema a)]
+            ;; Append to transaction log.
+            (<? (exec
+                  ["INSERT INTO transactions VALUES (?, ?, ?, ?, ?)" e a v tx (if added 1 0)]))
+            ;; Update materialized datom view.
+            (if (.-added datom)
+              (let [v (if fulltext?
+                        (<? (<insert-fulltext-value db v))
+                        v)]
+                (<? (exec
+                      ["INSERT INTO datoms VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)" e a v tx
+                       (ds/indexing? schema a) ;; index_avet
+                       (ds/ref? schema a) ;; index_vaet
+                       fulltext? ;; index_fulltext
+                       (ds/unique-value? schema a) ;; unique_value
+                       (ds/unique-identity? schema a) ;; unique_identity
```

Comment:
Probably worth memoizing all of them (at the schema level).  I need to extract `ident` and `entid` for real; this can fit into that refactoring.